File: 12、其余笔记/5、CSDN工程师认证/1、C1认证/4、作业四/作业四/3、自测题/1、冒泡排序/c代码/bubbleSort.c
#include <stdio.h>

//冒泡：升序排列
void bubble_sort(int array[], int length){

    for(int i = 0; i < length-1; i++){            //需要：确定n-1个位置， 第n个元素自然也就确定了
        int flag = 1;                            //flag为1时：默认为（已排序好）
        for(int j = 0; j < length-1-i; j++){    //需要：进行 n-1 - 已排序好的结点轮交换
            if(array[j] > array[j+1]){
                int temp = array[j];
                array[j] = array[j+1];
                array[j+1] = temp;
                flag = 0;                       //flag为0时：默认为（未排序好）
            }       
        }

        if(flag){
            break;                              //flag为1：说明已排序好，跳出循环即可
        }
    }
}

int main(void){
    int array[] = {0,9,12,1,6,3,7,11};  
    int length = sizeof(array) / sizeof(array[0]);  //计算：长度

    printf("排序前：");
    for(int i = 0; i < length; i++){
        printf("%d ", array[i]);
    }

    bubble_sort(array, length);             //调用排序：函数

    printf("\n排序后：");
    for(int i = 0; i < length; i++){
        printf("%d ", array[i]);
    }

}

File: 12、其余笔记/5、CSDN工程师认证/1、C1认证/4、作业四/作业四/3、自测题/1、冒泡排序/Java代码/BubbleSort.java
/**
 * Author: 老洋
 * Date:  2021/4/18 14:55
 */
package com.yyy;

import java.util.Arrays;

//实现对数组[0,9,12,1,6,3,7,11]的冒泡排序
public class BubbleSort {
    public static void main(String[] args){
        int[] oldArray = {0,9,12,1,6,3,7,11};

        System.out.println("旧数组：" + Arrays.toString(oldArray));
        System.out.println("新数组：" + Arrays.toString(bubbleSort(oldArray)));
    }

    public static int[] bubbleSort(int[] oldArray){
        //1、复制一份（旧数组）
        int[] newArray = Arrays.copyOf(oldArray, oldArray.length);

        //2、要进行：n-1次轮， 便可以确定下n个元素
        for(int i = 0; i < newArray.length-1; i++){
            boolean flag = true;            //默认为：序列已（完成排序）

            //3、每一轮：需要进行 n - 1 - i 次交换, 便可以（冒出来一个：元素）
            for(int j = 0; j < newArray.length-1-i; j++){
                //4、进行交换
                if(newArray[j] > newArray[j+1]){
                    int temp = newArray[j];
                    newArray[j] = newArray[j+1];
                    newArray[j+1] = temp;
                    flag = false;           //说明：序列（未完成排序）
                }
            }

            if(flag){
                break;                      //为true说明：（已完成排序）, 不用再进行遍历排序了
            }
        }
        return newArray;
    }

}